### ISSUE TITLE
Organizations and roles spec

### ISSUE THREAD

**Issue by MichaelJamesParsons** [unkso/titan-app#62]
## Description
Organizations and their corresponding roles must be migrated from static 12. The following outlines the spec for how the org/role relationship should be implemented.

#### Organization roles
| id | organization_id | user_id | role     | rank |
|----|-----------------|---------|----------|------|
| 1  | 9               | 3213    | CO       | 1    |
| 2  | 9               | 5435    | XO       | 2    |
| 3  | 9               | null    | NCOIC    | 3    |
| 4  | 9               | 9382    | Chaplain | null |

An `organization_role` represents a leadership position that may exist within an organization. Organizations in the community may have zero or more roles.

### Determining the authority of a role
The authority of each role is identified by the `rank` column. The order of authority between each role follows the conventions of a race:

```
1 = most authority, 2 = second most, 3 = third most, ...
```
#### Rules for ranks:
1. No organization may contain multiple roles with the same rank. Duplicates would make it difficult to derive a lower ranking member's CoC.
2. `null` ranks identify support roles that don't affect the CoC (i.e chaplain, assistants, etc.). These roles are not considered when calculating the CoC for members in child organizations.

### Identifying leadership
The `user_id` field indicates which member holds a given role.

1. Roles with a `null` user id indicates the role exists, but is not currently filled.
2. Role names are unique.
3. No user may have multiple roles in an organization.

### Splitting up duplicate roles
No organization may have multiple roles with the same name or rank. For instance, it doesn't make sense for a squad to have two NCO's, because it would become unclear which NCO is next in line for a member's CoC. If an organization requires duplicate positions, those positions should be distributed among child organizations:

> **Example:** The `navy` branch may need 3 NCO's: one for each squad. Instead of the navy organization having 3 NCO positions, it should have three child organizations; Each with a single NCO.

### User to organization
| id | organization_id | user_id |
|----|-----------------|---------|
| 1  | 28              | 3213    |
| 2  | 43              | 5435    |
| 3  | 19              | 1910    |

A user may be associated with 0-m organizations. Each row in `users_to_organizations` represents membership, not leadership. So users with NCO, CO, or any other leadership position will not be included in this table. 

Typically only "leaf organizations" (organizations with no children) have non-leadership members. It doesn't make sense for a user to be a member of the Navy or Intel. Rather, the user is generally a member of a Navy squad or Intel department. Therefore, the `user to organization` association should be limited between users and leaf organizations: squads, departments, JCS, teams, extracurricular groups, etc.

The representation of membership in non-leaf organizations (a division, general group) will be established through entries in the `organization_roles` table.

#### Get an organization
```
GET /organizations/<id>
```

> **Removed:** POST route is not required for current use cases.

#### Get/update an organization's roles
```
GET /organizations/<id>/roles
```
> **Removed:** POST|DELETE routes are not required for current use cases.

#### Get/add members to for an organization
The optional `children` parameter determines if members in child organizations should be included in the result set.
```
GET /organizations/<id>/members?children
```

> **Removed:** POST|DELETE routes are not required for current use cases.

#### Get member's CoC for a specific organization
```
GET /organizations/<id>/members/<id>/coc
```

#### Get all organizations a member belongs to.
The `type` parameter indicates which associations should be included (i.e only return organizations where the user has a role, no role, or either). 
```
GET /users/<id>/organizations?member=true&role=true
```

> **Updated:** Rocket doesn't support array type query strings. The `type` parameter has been replaced with two distinct parameters: `member: bool` and `role: bool`.

### Identifying a member's organizations
```rust
struct UserOrganizationMembership {
    pub organization_id: i32,
    pub role: Option<i32>  // If none, user does not hold a leadership role.
}
```
Using the example struct above:

1. Query `users_to_organizations` where `user_id = ?`.
2. Query `organization_roles` where `user_id = ?`.
3. Merge the two results into a vector of `UserOrganizationMembership` (the struct can change to fit more advanced needs if necessary).

### Query a user's CoC
1. Query the organization for the next CoC:
    1.1 If `role` is empty, query the user with the lowest ranking leadership position in the organization.
    1.2. If `role` is not empty, query the user with the next lowest leadership position in the organization.
2. If no user is found, recursively query parent organizations, repeating steps 1.1 and 1.2 until a user is found or there are now more results to search.

## Related use cases
1. Use CoC to determine who should receive reports for review.
2. Use CoC to create events that are mandatory for direct reports, but optional for all other members.
3. Derive permissions to perform CoC driven administrative actions.

**Comment by ELD** [comment]
@MichaelJamesParsons I'll review this proposal after we ship `0.1` and if I believe we need a system like this and the implementation looks good, I'll prioritize it for `0.1.5` or a future milestone. This isn't a necessity prior to shipping given our current barebones ACL implementation.

**Comment by ELD** [comment]
Right, I get that, but in `0.1`, we're not displaying units of finer resolution than a branch. If that wasn't clear ahead of time, I apologize, but displaying platoons, squads, etc was never the basic plan for `0.1`. The key functionality for `0.1` is to be able to add file entries for members and view their membership history. Future versions of Titan will display units at finer resolutions. Hence why the `branch` assumption is still valid at this point in time.

We're not implementing any of those use-cases in Titan `0.1` at this time. We can figure out which milestone these things fit into at a later date.

**Comment by MichaelJamesParsons** [comment]
We are rolling out `v0.1` without support for reports, attendance, or the ability to post excuses. These are key features that will block most NCO operations if not present. It may be in our best interest to delay a week so we can implement basic support for these features to avoid inevitable frustration.

**Comment by MichaelJamesParsons** [comment]
I recognize this addresses a problem without suggesting a solution. Here's my recommendation:

### Submitting excuses
The excuses system isn't associated with the events. It's completely stand-alone, which means we can import the excuses tables from static12 as-is, then implement a basic UI and backend to list and save excuses. Acknowledging an excuse will have to wait until we settle on a solution to determine a member's CoC.

### Attendance
Similarly, we can import the attendance data from static12, then implement basic support for indicating whether a user was `present`, `absent`, or `excused`.

### Reports
Reports is will require us to know a member's role & CoC in an organization. If we can go ahead and implement this spec now, we can also write a simple UI/backend for users to submit and view simple text reports.

## Long term
Once we have these fundamental features implemented, we can revisit them at a later time to develop more intuitive workflows. For now, a re-skin is our best option to quickly make these features available.

**Comment by ELD** [comment]
I appreciate the recommendations.

Here's the situation: we're not going to delay a week. We're going to ship what we have tomorrow and roll out more (such as data migration) this weekend. The branch selection UI will be done tonight along with a few other security fixes (such as not leaking password hashes of users to the client) and things. I'll work out deployment tomorrow afternoon and get things up and running.

Over the next week, we can implement simplified (or more polished versions) of reports, but in discussions with others, there's a large conversation about excuses and attendance that needs to happen before we build a solution that doesn't solve a real problem.

We can talk more about this tonight on Discord, if you'd like. Bottom line, we're not halting a release just because of those things being missing. The core, important functionality to branch and division levels are player files. Those are mostly feature complete (barring a few things missing) and we can press forward with it. We'll backfill and continuously ship as we can.

There are some process changes I'm going to make with the governance of the project moving forward as well. I'll go into more detail either tonight or later this weekend on it, as well.

**Comment by ELD** [comment]
Overall, this seems good. My only complaint/concern is how Rocket handles this:

`GET /users/<id>/organizations?type[]=member&type[]=role`

Would it be better to do something like?

`GET /users/<id>/organizations?type=[member, role]`

Regardless, I think the approach is sound. I wasn't convinced of...

> No organization may have multiple roles with the same name or rank. For instance, it doesn't make sense for a squad to have two NCO's, because it would become unclear which NCO is next in line for a member's CoC. If an organization requires duplicate positions, those positions should be distributed among child organizations:

... until I read further. It makes sense. There's no precedent for co-leaders of a unit, so it makes sense to lock it down to one member per role.

Will this inform the drawing of a roster? If so, how will titles (i.e. "Boat Team Leader" versus "Squad Leader") factor into this proposal?

**Comment by MichaelJamesParsons** [comment]
I forgot rocket won't have support for array type query string parameters until `v.5`. We can pass the parameter as a string using a comma delimiter:

```
GET /users/<id>/organizations?type=member,role
```

> There's no precedent for co-leaders of a unit, so it makes sense to lock it down to one member per role.

If we eventually need co-leaders or assistant NCO's, we can assign a `null` rank to the secondary leader. This maintains a clear CoC and allows for duplicates of the same role in an organization.

> Will this inform the drawing of a roster? If so, how will titles (i.e. "Boat Team Leader" versus "Squad Leader") factor into this proposal?

Roles aren't restricted to any name convention. Names can be tailored to the organization. Alternatively, we can add an `alias` column to the roles table so we can reference the position as `squad leader` or `boat team leader` where appropriate.
